internal/names/names.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2025 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Forked from https://github.com/knative/pkg/blob/ee1db869c7ef25eb4ac5c9ba0ab73fdc3f1b9dfa/kmeta/names.go
+
+/*
+copyright 2019 the knative authors
+
+licensed under the apache license, version 2.0 (the "license");
+you may not use this file except in compliance with the license.
+you may obtain a copy of the license at
+
+    http://www.apache.org/licenses/license-2.0
+
+unless required by applicable law or agreed to in writing, software
+distributed under the license is distributed on an "as is" basis,
+without warranties or conditions of any kind, either express or implied.
+see the license for the specific language governing permissions and
+limitations under the license.
+*/
+
+package names
+
+import (
+	"crypto/md5" //nolint:gosec // No strong cryptography needed.
+	"encoding/hex"
+	"fmt"
+	"regexp"
+)
+
+// The longest name supported by the K8s is 63.
+// These constants.
+const (
+	longest = 63
+	md5Len  = 32
+	head    = longest - md5Len // How much to truncate to fit the hash.
+)
+
+var isAlphanumeric = regexp.MustCompile(`^[a-zA-Z0-9]*$`)
+
+// ChildName generates a name for the resource based upon the parent resource and suffix.
+// If the concatenated name is longer than K8s permits the name is hashed and truncated to permit
+// construction of the resource, but still keeps it unique.
+// If the suffix itself is longer than 31 characters, then the whole string will be hashed
+// and `parent|hash|suffix` will be returned, where parent and suffix will be trimmed to
+// fit (prefix of parent at most of length 31, and prefix of suffix at most length 30).
+func ChildName(parent, suffix string) string {
```

Comment:
i did not super scrutinize the logic in this function, definitely leaning a bit on the years of successful usage in knative 🙏